File: src/java/org/kowboy/SpittoonPlugin.java
package org.kowboy;

import org.bukkit.plugin.java.JavaPlugin;
import clojure.lang.IFn;
import clojure.java.api.Clojure;

public class SpittoonPlugin extends JavaPlugin {
    @Override
    public void onEnable() {
        ClassLoader previous = Thread.currentThread().getContextClassLoader();
        try {
            ClassLoader cl = getClassLoader();
            Thread.currentThread().setContextClassLoader(cl);

            IFn require = Clojure.var("clojure.core", "require");
            require.invoke(Clojure.read("org.kowboy.bukkit.plugin"));
            require.invoke(Clojure.read("org.kowboy.bukkit.repl"));

            IFn enable = Clojure.var("org.kowboy.bukkit.plugin", "enable");
            enable.invoke(this);
        } finally {
            Thread.currentThread().setContextClassLoader(previous);
        }
    }

    @Override
    public void onDisable() {
        ClassLoader previous = Thread.currentThread().getContextClassLoader();
        try {
            ClassLoader cl = getClassLoader();
            Thread.currentThread().setContextClassLoader(cl);

            IFn disable = Clojure.var("org.kowboy.bukkit.plugin", "disable");
            disable.invoke(this);
        } finally {
            Thread.currentThread().setContextClassLoader(previous);
        }
    }

    public static void main(String[] args) {
        IFn plus = Clojure.var("clojure.core", "+");
        System.out.println(plus.invoke(1, 2));
    }
}

File: src/main/kotlin/org/kowboy/SpittoonPlugin.kt
package org.kowboy

import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import org.kowboy.command.SpitCommandExecutor
import org.kowboy.listener.LightLevelListener

/**
 *
 *
 * A minecraft server plugin which implements a random assortment of arguably useful features. This collection of
 * junk is fondly referred to as "Spittoon" (a receptacle for tobacco spit...yum).
 *
 *
 *
 *
 * Current features implemented:
 * <dl>
 * <dt>light-level</dt>
 * <dd>
 * When enabled, displays indicators of light-level on the top face of blocks near the player.
 * Useful for lighting up an area to prevent mob spawning. See [LightLevelListener] for more detail.
</dd> *
 * <dt>find</dt>
 * <dd>
 * Find nearby blocks, entities, and slime chunks. See [FindCommandExecutor] for more detail.
</dd> *
 * <dt>journal</dt>
 * <dd>
 * Keep a player journal with teleport bookmarks. See [JournalCommandExecutor] for more detail.
</dd> *
</dl> *
 *
 *
 * @author Craig McDaniel
 * @since 1.0
 */
class SpittoonPlugin : JavaPlugin() {
    override fun onEnable() {
        // Be sure to write out plugins/Spittoon/config.yaml
        // Does not overwrite if it already exists.
        saveDefaultConfig()
        // Commands
        addCommands()
        // Listeners
        addListeners()
    }

    private fun addListeners() {
        LightLevelListener(this)
    }

    /**
     * Convenience callback for event listeners to register themselves in their constructor.
     * @param listener the listener to register events for.
     */
    fun registerEvents(listener: Listener) {
        server.pluginManager.registerEvents(listener, this)
    }

    private fun addCommands() {
        SpitCommandExecutor(this)
    }

    override fun onDisable() {
        saveConfig()
        super.onDisable()
    }
}